File: tests/Unit/DataStructures/Tensor/Test_TensorData.cpp
// Distributed under the MIT License.
// See LICENSE.txt for details.

#include "Framework/TestingFramework.hpp"

#include <cstddef>
#include <string>
#include <vector>

#include "DataStructures/DataVector.hpp"
#include "DataStructures/Tensor/TensorData.hpp"
#include "Framework/TestHelpers.hpp"
#include "Utilities/GetOutput.hpp"
#include "Utilities/StdHelpers.hpp"  // IWYU pragma: keep

namespace {
template <typename DataType>
void test() {
  TensorComponent tc0("T_x", DataType{8.9, 7.6, -3.4, 9.0});
  TensorComponent tc1("T_y", DataType{8.9, 7.6, -3.4, 9.0});
  TensorComponent tc2("T_x", DataType{8.9, 7.6, -3.4, 9.1});
  CHECK(get_output(tc0) == "(T_x, (8.9,7.6,-3.4,9))");
  CHECK(tc0 == tc0);
  CHECK(tc0 != tc1);
  CHECK(tc0 != tc2);
  CHECK(tc1 != tc2);
  test_serialization(tc0);

  const ExtentsAndTensorVolumeData etvd0({2, 2}, {tc0, tc1, tc2});
  const auto after = serialize_and_deserialize(etvd0);
  CHECK(after.extents == etvd0.extents);
  CHECK(after.tensor_components == etvd0.tensor_components);
  CHECK(after == etvd0);
  // Check operator==
  CHECK(etvd0 != ExtentsAndTensorVolumeData({3, 2}, {tc0, tc1, tc2}));
  CHECK(etvd0 != ExtentsAndTensorVolumeData({2, 2}, {tc2, tc1, tc0}));
  CHECK(etvd0 == ExtentsAndTensorVolumeData({2, 2}, {tc0, tc1, tc2}));

  const ElementVolumeData evd0(
      {2, 2}, {tc0, tc1, tc2},
      {Spectral::Basis::Legendre, Spectral::Basis::Chebyshev},
      {Spectral::Quadrature::Gauss, Spectral::Quadrature::GaussLobatto},
      "Element0");
  const auto after_evd0 = serialize_and_deserialize(evd0);
  CHECK(after_evd0.extents == evd0.extents);
  CHECK(after_evd0.tensor_components == evd0.tensor_components);
  CHECK(after_evd0.basis == evd0.basis);
  CHECK(after_evd0.quadrature == evd0.quadrature);
  CHECK(after_evd0.element_name == evd0.element_name);
  CHECK(after_evd0 == evd0);
  // Check operator==
  CHECK(evd0 !=
        ElementVolumeData(
            {3, 2}, {tc0, tc1, tc2},
            {Spectral::Basis::Legendre, Spectral::Basis::Chebyshev},
            {Spectral::Quadrature::Gauss, Spectral::Quadrature::GaussLobatto},
            "Element0"));
  CHECK(evd0 !=
        ElementVolumeData(
            {2, 2}, {tc2, tc1, tc0},
            {Spectral::Basis::Legendre, Spectral::Basis::Chebyshev},
            {Spectral::Quadrature::Gauss, Spectral::Quadrature::GaussLobatto},
            "Element0"));
  CHECK(evd0 !=
        ElementVolumeData(
            {2, 2}, {tc0, tc1, tc2},
            {Spectral::Basis::Chebyshev, Spectral::Basis::Chebyshev},
            {Spectral::Quadrature::Gauss, Spectral::Quadrature::GaussLobatto},
            "Element0"));
  CHECK(evd0 != ElementVolumeData(
                    {2, 2}, {tc0, tc1, tc2},
                    {Spectral::Basis::Legendre, Spectral::Basis::Chebyshev},
                    {Spectral::Quadrature::GaussLobatto,
                     Spectral::Quadrature::GaussLobatto},
                    "Element0"));
  CHECK(evd0 !=
        ElementVolumeData(
            {2, 2}, {tc0, tc1, tc2},
            {Spectral::Basis::Legendre, Spectral::Basis::Chebyshev},
            {Spectral::Quadrature::Gauss, Spectral::Quadrature::GaussLobatto},
            "Element1"));
  CHECK(evd0 ==
        ElementVolumeData(
            {2, 2}, {tc0, tc1, tc2},
            {Spectral::Basis::Legendre, Spectral::Basis::Chebyshev},
            {Spectral::Quadrature::Gauss, Spectral::Quadrature::GaussLobatto},
            "Element0"));
}
}  // namespace

SPECTRE_TEST_CASE("Unit.DataStructures.TensorData", "[Unit]") {
  test<DataVector>();
  test<std::vector<float>>();
}

File: tests/Unit/DataStructures/Tensor/Test_TensorData.py
# Distributed under the MIT License.
# See LICENSE.txt for details.

from spectre.DataStructures import (DataVector, ExtentsAndTensorVolumeData,
                                    TensorComponent, Legendre, Gauss,
                                    ElementVolumeData)

import unittest
import numpy as np
import numpy.testing as npt


class TestTensorData(unittest.TestCase):
    # Tests for TensorComponent functions
    def test_tensor_component(self):
        # Set up Tensor Component
        tensor_component = TensorComponent("tensor component",
                                           DataVector([1.5, 1.1]))
        # Test name
        self.assertEqual(tensor_component.name, "tensor component")
        tensor_component.name = "new tensor component"
        self.assertEqual(tensor_component.name, "new tensor component")
        # Test data
        npt.assert_array_almost_equal(np.array(tensor_component.data),
                                      np.array([1.5, 1.1]))
        tensor_component.data = DataVector([6.7, 3.2])
        npt.assert_array_almost_equal(np.array(tensor_component.data),
                                      np.array([6.7, 3.2]))
        # Test str, repr
        self.assertEqual(str(tensor_component),
                         "(new tensor component, (6.7,3.2))")
        self.assertEqual(repr(tensor_component),
                         "(new tensor component, (6.7,3.2))")

    # Tests for ExtentsAndTensorVolumeData functions
    def test_extents_and_tensor_volume_data(self):
        # Set up Extents and Tensor Volume data
        tensor_component_1 = TensorComponent("tensor component one",
                                             DataVector([1.5, 1.1]))
        tensor_component_2 = TensorComponent("tensor component two",
                                             DataVector([7.1, 5]))
        extents_and_data = ExtentsAndTensorVolumeData(
            [1, 2, 3, 4], [tensor_component_1, tensor_component_2])
        # Test extents
        self.assertEqual(extents_and_data.extents, [1, 2, 3, 4])
        extents_and_data.extents = [5, 6, 7, 8]
        self.assertEqual(extents_and_data.extents, [5, 6, 7, 8])
        # Test tensor components
        self.assertEqual(extents_and_data.tensor_components,
                         [tensor_component_1, tensor_component_2])
        extents_and_data.tensor_components = [
            tensor_component_1, tensor_component_1
        ]
        self.assertEqual(extents_and_data.tensor_components,
                         [tensor_component_1, tensor_component_1])
        # Test str, repr
        self.assertEqual(
            str(extents_and_data),
            "((5,6,7,8),((tensor component one, (1.5,1.1)"
            "),(tensor component one, (1.5,1.1))))")
        self.assertEqual(
            repr(extents_and_data),
            "((5,6,7,8),((tensor component one, (1.5,1.1)"
            "),(tensor component one, (1.5,1.1))))")

        # Tests for ExtentsAndTensorVolumeData functions
    def test_element_volume_data(self):
        # Set up Extents and Tensor Volume data
        tensor_component_1 = TensorComponent("tensor component one",
                                             DataVector([1.5, 1.1]))
        tensor_component_2 = TensorComponent("tensor component two",
                                             DataVector([7.1, 5]))
        basis = Legendre
        quad = Gauss
        element_data = ElementVolumeData(
            [1, 2, 3, 4], [tensor_component_1, tensor_component_2],
            [basis, basis], [quad, quad])

        # Test extents
        self.assertEqual(element_data.extents, [1, 2, 3, 4])
        element_data.extents = [5, 6, 7, 8]
        self.assertEqual(element_data.extents, [5, 6, 7, 8])
        # Test tensor components
        self.assertEqual(element_data.tensor_components,
                         [tensor_component_1, tensor_component_2])
        element_data.tensor_components = [
            tensor_component_1, tensor_component_1
        ]
        self.assertEqual(element_data.tensor_components,
                         [tensor_component_1, tensor_component_1])
        # Test str, repr
        self.assertEqual(
            str(element_data), "((5,6,7,8),((tensor component one, (1.5,1.1)"
            "),(tensor component one, (1.5,1.1))))")
        self.assertEqual(
            repr(element_data), "((5,6,7,8),((tensor component one, (1.5,1.1)"
            "),(tensor component one, (1.5,1.1))))")
        # Test basis and quadrature
        self.assertEqual(element_data.basis, [basis, basis])
        self.assertEqual(element_data.quadrature, [quad, quad])


if __name__ == '__main__':
    unittest.main(verbosity=2)
